scilifelab_parsers/qc/qc.py

```diff
@@ -129,78 +129,34 @@ def by_lane_yield(data):
-        if not smp_header == smp_header_known:
-            self.log.warn("Sample header information has changed. New "
-                    "format?\nOld format: {0}\nSaw: {1}".format(
-                    ",".join((["'{0}'".format(x) for x in smp_header_known])),
-                    ",".join(["'{0}'".format(x) for x in smp_header])))
-        ## Fix first header name in smp_header since htm document is mal-formatted: <th>Sample<p></p>ID</th>
-        smp_header[0] = "Sample ID"
-
-        ## Parse Barcode lane statistics
-        soup = BeautifulSoup(htm_doc)
-        table = soup.findAll("table")[1]
-        rows = table.findAll("tr")
-        column_gen = (row.findAll("td") for row in rows)
-        parse_row = lambda row: dict([(bc_header[i],str(row[i].string)) for i in range(0, len(bc_header)) if row])
-        metrics["Barcode_lane_statistics"].extend(map(parse_row, column_gen))
-
-        ## Parse Sample information
-        soup = BeautifulSoup(htm_doc)
-        table = soup.findAll("table")[3]
-        rows = table.findAll("tr")
-        column_gen = (row.findAll("td") for row in rows)
-        parse_row = lambda row: dict([(smp_header[i],str(row[i].string)) for i in range(0, len(smp_header)) if row])
-        metrics["Sample_information"].extend(map(parse_row, column_gen))
-
-        # Define a function for sorting the values
-        def by_lane_sample(data):
-            return "{0}-{1}-{2}".format(data.get('Lane',''),data.get('Sample ID',''),data.get('Index',''))
-
-        # Post-process the metrics data to eliminate duplicates resulting from multiple stats files
-        for metric in ['Barcode_lane_statistics', 'Sample_information']:
-            dedupped = {}
-            for row in metrics[metric]:
-                key = "\t".join(row.values())
-                if key not in dedupped:
-                    dedupped[key] = row
-                else:
-                    self.log.debug("Duplicates of Demultiplex Stats entries discarded: {0}".format(key[0:min(35,len(key))]))
-            metrics[metric] = sorted(dedupped.values(), key=by_lane_sample)
-
-        ## Set data
+    def parse_demultiplex_stats_htm(self, html_file):
+        metrics = self._html_tables_to_lists_of_tuples(html_file)    
+        new_metrics = {}
+        for header, table in metrics.items():
+            new_metrics[header] = []
+            for sample in table:
+                new_metrics[header].append(dict(sample))
+        return new_metrics
+
+    def _html_tables_to_lists_of_tuples(self, html):
+        html = open(html,'r')
+        bs = BeautifulSoup(html)
+        metrics = {}
+        keys = []
+        header = None
+        for table in bs.findAll('table'):
+            header = table.findPrevious('h2')
+            if header:
+                header = header.text.replace(' ','_')
+            if header not in metrics.keys():
+                metrics[header] = []
+            if not metrics[header]: 
```

Comment:
cat we merge both if loops ? I mean if this `if not metrics[header]` is necessary at all, as there is no any `else` block ?